### Fitting Logistic Regression

In this first notebook, you will be fitting a logistic regression model to a dataset where we would like to predict if a transaction is fraud or not.

To get started let's read in the libraries and take a quick look at the dataset.

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm


df = pd.read_csv('./fraud_dataset.csv')
df.head()

transaction_id   duration      day  fraud
0           28891  21.302600  weekend  False
1           61629  22.932765  weekend  False
2           53707  32.694992  weekday  False
3           47812  32.784252  weekend  False
4           43455  17.756828  weekend  False

`1.` As you can see, there are two columns that need to be changed to dummy variables.  Replace each of the current columns to the dummy version.  Use the 1 for `weekday` and `True`, and 0 otherwise.  Use the first quiz to answer a few questions about the dataset.

In [2]:
df['weekday'] = pd.get_dummies(df['day'])['weekday']
df[['not_fraud','fraud']] = pd.get_dummies(df['fraud'])
df = df.drop('not_fraud', axis=1)

In [4]:
df.head()

transaction_id   duration      day  fraud  weekday
0           28891  21.302600  weekend      0        0
1           61629  22.932765  weekend      0        0
2           53707  32.694992  weekday      0        1
3           47812  32.784252  weekend      0        0
4           43455  17.756828  weekend      0        0

In [3]:
print(df['fraud'].mean())
print(df['weekday'].mean())
print(df.groupby('fraud').mean()['duration'])

0.012168770612987604
0.3452746502900034
fraud
0    30.013583
1     4.624247
Name: duration, dtype: float64


- The proportion of fraudulent transactions. 0.012


- The average duration for fraudulent transaction. 4.62


- The proportion of weekday transactions. 0.345


- The average duration for non-fraudulent transactions. 30.01




`2.` Now that you have dummy variables, fit a logistic regression model to predict if a transaction is fraud using both day and duration.  Don't forget an intercept!  Use the second quiz below to assure you fit the model correctly.

In [11]:
df['intercept'] = 1
log_mod = sm.Logit(df['fraud'], df[['intercept', 'weekday', 'duration']])
results = log_mod.fit()
results.summary()

Optimization terminated successfully.
         Current function value: inf
         Iterations 16


/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1214: RuntimeWarning: overflow encountered in exp
  return 1/(1+np.exp(-X))
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1264: RuntimeWarning: divide by zero encountered in log
  return np.sum(np.log(self.cdf(q*np.dot(X,params))))
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:2397: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - self.llf/self.llnull
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:2401: RuntimeWarning: invalid value encountered in double_scalars
  return -2*(self.llnull - self.llf)


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  fraud   No. Observations:                 8793
Model:                          Logit   Df Residuals:                     8790
Method:                           MLE   Df Model:                            2
Date:                Sun, 26 Nov 2017   Pseudo R-squ.:                     nan
Time:                        22:05:17   Log-Likelihood:                   -inf
converged:                       True   LL-Null:                          -inf
                                        LLR p-value:                       nan
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
intercept      9.8709      1.944      5.078      0.000       6.061      13.681
weekday        2.5465      0.904      2.816      0.005       0.774       4.319
duration      -1.4637      0.290     -5.039      0.000      -2.033      -0.894
==============================================================================

Possibly complete quasi-separation: A fraction 0.98 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

In [6]:
np.exp(2.5465)

12.762357271496972

In [7]:
np.exp(-1.4637) 
# <1, should count reciprocal, like below

0.2313785882117941

In [8]:
1/np.exp(-1.4637)

4.321921089278333

Which variables are statistically significant for predicting fraud? Check all that apply.

- Duration.

- Weekday.


Which of the following interpretations are correct with regard to your coefficients? Check all that apply.

- On weekdays, the chance of fraud is 12.76 times more likely than on weekends holding duration constant. 
- For each minute less spent on the transaction, the chance of fraud is 4.32 times more likely holding the day of week constant.
